plugins/SpectrumAnalyzer/SaProcessor.h

```diff
@@ -0,0 +1,119 @@
+/* SaProcessor.h - declaration of SaProcessor class.
+ *
+ * Copyright (c) 2019 Martin Pavelek <he29/dot/HS/at/gmail/dot/com>
+ *
+ * Based partially on Eq plugin code,
+ * Copyright (c) 2014 David French <dave/dot/french3/at/googlemail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef SAPROCESSOR_H
+#define SAPROCESSOR_H
+
+#include <mutex>
+#include <QImage>
+#include <vector>
+
+#include "fft_helpers.h"
+#include "SaControls.h"
+
+
+// Receives audio data, runs FFT analysis and stores the result.
+class SaProcessor {
+public:
+	SaProcessor(SaControls *controls);
+	virtual ~SaProcessor();
+
+	void analyse(sampleFrame *in_buffer, const fpp_t frame_count);
+
+	// inform processor if any processing is actually required
+	void setSpectrumActive(bool active);
+	void setWaterfallActive(bool active);
+
+	// configuration is taken from models in SaControls; some changes require
+	// an exlicit update request (reallocation and window rebuild)
+	void reallocateBuffers();
+	void rebuildWindow();
+	void clear();
+
+	// information about results and unit conversion helpers
+	float binToFreq(int index);
+	float binBandwidth();
+
+	float freqToXPixel(float frequency, int width);
+	float xPixelToFreq(float x, int width);
+
+	float ampToYPixel(float amplitude, int height);
+	float yPixelToAmp(float y, int height);
+
+	int getSampleRate() const;
+	float getFreqRangeMin(bool linear = false);
+	float getFreqRangeMax();
+	float getAmpRangeMin(bool linear = false);
+	float getAmpRangeMax();
+
+	// data access lock must be acquired by any friendly class that touches
+	// the results, mainly to prevent unexpected mid-way reallocation
+	std::mutex m_dataAccess;
```

Comment:
`std::mutex` is unusable on MinGW compilers with Win32 threading model. You can use `QMutex` from Qt instead.
Also, Please consider using lock guards (`std::lock_guard` for `std::mutex`, `QMutexLocker` for `QMutex`) instead of direct locks/unlocks.

Reply:
Thanks, I'll look into it. I don't have much prior experience with multi threaded applications, so I did not even know about the lock guard concept. Would have saved me from some dead-locks during development... :)